crates/typst-syntax/src/parser.rs

```diff
@@ -1494,23 +1497,70 @@ fn pattern_leaf<'s>(
     }
 }
 
-/// Manages parsing of a stream of tokens.
+/// A single parsed Token.
+#[derive(Clone)]
+struct Token<'s> {
+    kind: SyntaxKind,
+    text: &'s str,
+}
+
+/// The starting point of any trivia (whitespace/comments) preceding the current token.
+#[derive(Clone)]
+struct TriviaStart {
+    /// Number of preceding trivia `nodes` (for avoiding trivia when wrapping).
+    num: usize,
+    /// Offset into `text` of the trivia start (for moving the lexer's cursor back).
+    offset: usize,
+}
+
+/// Manages parsing of a stream of tokens into a tree of SyntaxNodes.
+///
+/// This acts like a bottom-up (shift-reduce) parser, where we investigate a current token
+/// and either:
+/// 1. Eat (shift) another token, putting `current` into the `nodes` vector.
+/// 2. Wrap (reduce) the nodes up to a marker (excluding `current`) into a corresponding
+///    SyntaxKind.
+///
+/// In the end we have a nested tree of SyntaxNodes as a Concrete Syntax Tree. This will
+/// have all irrelevant tokens (whitespace, comments, code parens, commas in function
+/// args, etc) removed when interpreted as an Abstract Syntax Tree for evaluation. The
```

Comment:
"removed" here sounds like the AST is built eagerly rather than being a lazy view in top of the CST

Reply:
Agreed, addressed in the rewrite in the other comment.